backend/models/sequence.js

```diff
@@ -19,16 +19,102 @@
 const db = require("../handler/db");
 const responseCodes = require("../response_codes.js");
 const utils = require("../utils");
+const yup = require("yup");
 
+const { update: updateGroup } = require("./group");
 const History = require("./history");
 const FileRef = require("./fileRef");
 const { getRefNodes } = require("./ref");
 const { getDefaultLegendId } = require("./modelSetting");
+const View = new (require("./view"))();
+const { cleanViewpoint, createViewpoint } = require("./viewpoint");
 
-const activityCol = (modelId) => `${modelId}.activities`;
 const sequenceCol = (modelId) => `${modelId}.sequences`;
 const legendCol = (modelId) => `${modelId}.sequences.legends`;
 
+const findMinMaxDate =  (frames) => {
+	let min = frames[0].dateTime;
+	let max = frames[0].dateTime;
+
+	frames.forEach((frame) => {
+		if (frame.dateTime < min) {
+			min = frame.dateTime;
+		}
+		if (frame.dateTime > max) {
+			max = frame.dateTime;
+		}
+	});
+
+	return {min, max};
+};
+
+const byDateTime = (frameA, frameB) => frameA.dateTime - frameB.dateTime;
+
+const notEmpty = (obj) => Object.keys(obj).length > 0;
+
+// Viewpoint checks
+const viewpointSchema = yup.object().test((viewpoint) => {
+	if (!viewpoint) {
+		return true;
+	}
+
+	const {transformation_group_ids, transformation_groups } = viewpoint;
+
+	// The viewpoints for sequences CANT have transformations groups
+	if (transformation_group_ids || transformation_groups) {
+		return false;
+	}
+
+	return true;
+});
+
+const frameSchema = yup.object().shape({
+	dateTime: yup.number().required(),
+	viewpoint: viewpointSchema,
+	viewId: yup.string()
+}).noUnknown().test(({viewId, viewpoint}) => {
+	// The frame must have either viewpoint or viewId, not both
+	return (viewId && !viewpoint) || (!viewId && viewpoint);
+});
+
+const nameSchema = yup.string().min(1).max(30);
+
+const sequenceSchema = yup.object().shape({
+	name: nameSchema.required(),
+	startDate: yup.number(),
+	endDate: yup.number(),
+	rev_id: yup.string(),
+	frames: yup.array().of(frameSchema).min(1).required()
+}).noUnknown().test(({startDate, endDate, frames}) => {
+	if ((startDate !== undefined || endDate !== undefined) && frames) {
+		const {min, max} = findMinMaxDate(frames);
```

Comment:
2 questions here:

- we're looping through this to find min/max dates, later on in the code we have to sort the dates by order. Is it possible for yup to sort it here and find the minMax after that?
- If we can't do the above, maybe it's better to just leave this logic as a check manually?

In essence I want to avoid the double looping if possible

Reply:
We can do the first thing. I'll do the second one.